test/surface_test.py

```diff
@@ -815,41 +814,81 @@ def test_blit__SRCALPHA_opaque_source(self):
 
-        s = pygame.Surface(size, pygame.SRCALPHA, 32)
-        s2 = s.copy()
 
-        s.fill((32, 32, 32, 111))
-        s2.fill((32, 32, 32, 31))
+    def test_blit__SRCALPHA_to_SRCALPHA_non_zero(self):
+        """Tests blitting a nonzero alpha surface to another nonzero alpha surface
+         both premultiplied and straight alpha compositing methods."""
 
-        s.blit(s2, (0, 0))
+        (w, h) = size = (32, 32)
 
-        # TODO:
-        # what is the correct behaviour ?? should it blend? what algorithm?
+        def check_color_diff(color1, color2):
+            """Returns True if two colors are within (1, 1, 1, 1)"""
+            for val in (color1 - color2):
+                if abs(val) > 1:
+                    return False
+            return True
 
-        self.assertEqual(s.get_at((0, 0)), (32, 32, 32, 31))
+        def high_a_onto_low(high, low):
+            high_alpha_surface = pygame.Surface(size, pygame.SRCALPHA, 32)
+            low_alpha_surface = high_alpha_surface.copy()
+            high_alpha_color = Color((32, 32, 32, high))
+            low_alpha_color = Color((32, 32, 32, low))
+            high_alpha_surface.fill(high_alpha_color)
+            low_alpha_surface.fill(low_alpha_color)
+    
+            # Premultiplied case, src = low alpha, dst = high alpha
+            high_alpha_surface.blit(low_alpha_surface, (0, 0), special_flags=BLEND_PREMULTIPLIED)
+            
+            expected_color = (low_alpha_color) + Color(tuple(((x*(255-low_alpha_color.a))//255) for x in high_alpha_color))
+            self.assertTrue(check_color_diff(high_alpha_surface.get_at((0, 0)), expected_color))
+        
+        def low_a_onto_high(high, low):
+            high_alpha_surface = pygame.Surface(size, pygame.SRCALPHA, 32)
+            low_alpha_surface = high_alpha_surface.copy()
+            high_alpha_color = Color((32, 32, 32, high))
+            low_alpha_color = Color((32, 32, 32, low))
+            high_alpha_surface.fill(high_alpha_color)
+            low_alpha_surface.fill(low_alpha_color)
+    
+            # Premultiplied case, src = high alpha, dst = low alpha
+            low_alpha_surface.blit(high_alpha_surface, (0, 0), special_flags=BLEND_PREMULTIPLIED)
+            
+            expected_color = (high_alpha_color) + Color(tuple(((x*(255-high_alpha_color.a))//255) for x in low_alpha_color))
+            self.assertTrue(check_color_diff(low_alpha_surface.get_at((0, 0)), expected_color))
+        
+        # Test premultiplication on array of alpha values #TODO Straight Alpha Compositing
+        for low_a in range(0,128):
+            for high_a in range(128, 256):
+                high_a_onto_low(high_a, low_a)
+                low_a_onto_high(high_a, low_a)
+                
 
     def test_blit__SRCALPHA32_to_8(self):
         # Bug: fatal
         # SDL_DisplayConvert segfaults when video is uninitialized.
         target = pygame.Surface((11, 8), 0, 8)
-        color = target.get_palette_at(2)
```

Comment:
This was overwriting something in another namespace -> renamed.

Reply:
👍 thumbs